subsys/nrf_rpc/nrf_rpc_uart.c

```diff
@@ -176,39 +184,49 @@ static bool crc_compare(uint16_t rx_crc, uint16_t calc_crc)
 	return rx_crc == calc_crc;
 }
 
-static void hdlc_decode_byte(struct nrf_rpc_uart *uart_tr, uint8_t byte)
+static void hdlc_decode_byte(enum hdlc_state *state, uint8_t *out, size_t *len, size_t capacity,
+			     uint8_t byte)
 {
-	/* TODO: handle frame buffer overflow */
-
-	if (uart_tr->hdlc_state == HDLC_STATE_ESCAPE) {
-		uart_tr->rx_packet[uart_tr->rx_packet_len++] = byte ^ 0x20;
-		uart_tr->hdlc_state = HDLC_STATE_FRAME_START;
-	} else if (byte == HDLC_CHAR_ESCAPE) {
-		uart_tr->hdlc_state = HDLC_STATE_ESCAPE;
-	} else if (byte == HDLC_CHAR_DELIMITER) {
-		uart_tr->hdlc_state = (uart_tr->hdlc_state == HDLC_STATE_FRAME_START)
-					      ? HDLC_STATE_FRAME_FOUND
-					      : HDLC_STATE_UNSYNC;
-	} else {
-		uart_tr->rx_packet[uart_tr->rx_packet_len++] = byte;
-		uart_tr->hdlc_state = HDLC_STATE_FRAME_START;
+	switch (*state) {
+	case HDLC_STATE_UNSYNC:
+		if (byte == HDLC_CHAR_DELIMITER) {
+			*len = 0;
+			*state = HDLC_STATE_FRAME;
+		}
+		return;
+	case HDLC_STATE_FRAME_FOUND:
+		*len = 0;
+		*state = HDLC_STATE_FRAME;
+		__fallthrough;
+	case HDLC_STATE_FRAME:
+		if (byte == HDLC_CHAR_DELIMITER) {
+			if (*len > 0) {
```

Comment:
what if delimiter is found without prior data? is this empty frame or protocol does not allow this?

Reply:
HDLC defines that the delimiter should be present at the start and end of the frame but if 2 frames are sent back to back you can either send two delimiters (end of the previous frame and start of the next frame) or just one. I think this implies that empty frames cannot be supported.